In [1]:
import csv
import numpy as np
from sklearn.cross_validation import KFold
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.naive_bayes import GaussianNB
from sklearn.linear_model import LinearRegression
from sklearn.svm import LinearSVC


C:\Users\ar1\Anaconda3\lib\site-packages\sklearn\cross_validation.py:44: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


In [2]:
def findCommunities():
    communities = []
    with open('C:\\Users\\ar1\\Documents\ML\\Project\\Crime Prediction Data(1)\\Crime Prediction Data\\communities-crime-clean.csv', 'r') as csvfile:
        filereader = csv.reader(csvfile)
        j = 0
        for i in filereader:
            if j > 0:
                if i[1] not in communities:
                    communities.append(i[1])
            j+=1
    return communities

In [3]:
def extractData(filename,communities,addstate = False,addcommunity = False):
    features = []
    x=[]
    y=[]
    with open(filename, 'r') as csvfile:
        filereader = csv.reader(csvfile)
        j = 0
        for i in filereader:
            if j > 0:
                data = i[3:len(i)-1]
                if addstate:
                    state = [0.0]*56
                    state[int(i[0])-1] = 1.0
                    data = data + state
                    features = features + ["state"+str(i) for i in range(56)]
                if addcommunity:
                    tmp = [0.0]*len(communities)
                    tmp[communities.index(i[1])] = 1.0
                    data = data + tmp
                    features = features + communities
                vect = np.array(data)
                #vect = np.array(i[3:len(i)-1])
                x.append(vect.astype(np.float))
                if float(i[-1]) > 0.1:
                    y.append(1)
                else:
                    y.append(0)
            else:
                features = i[3:len(i)-1]
            j+=1
        y = np.array(y)
        x = np.array(x)
        features = np.array(features)
        return x,y,features


In [4]:
def meanSquaredError(true,predict):
    print("Mean squared error: %.4f"
      % np.mean((predict-true) ** 2))

In [5]:
def linearRegression():
    return LinearRegression()


In [6]:
def do_cross_validation(X, y,clf, n_folds=5):
    cv = KFold(len(y), n_folds)
    accuracies = []
    for train_ind, test_ind in cv: 
        clf.fit(X[train_ind], y[train_ind])
        predictions = clf.predict(X[test_ind])
        meanSquaredError(y[test_ind],predictions)

In [7]:
distintCommunities = findCommunities()

In [8]:
filename= 'C:\\Users\\ar1\\Documents\ML\\Project\\Crime Prediction Data(1)\\Crime Prediction Data\\communities-crime-clean.csv'
X,Y,Features = extractData(filename, distintCommunities)

In [9]:
LinReg= linearRegression()
LinReg.fit(X[:1000],Y[:1000])
predictedY=LinReg.predict(X[1000:])

In [10]:
meanSquaredError(predictedY,Y[1000:])

Mean squared error: 0.1603


In [11]:
do_cross_validation(X[:1000],Y[:1000],LinReg,10)

Mean squared error: 0.1883
Mean squared error: 0.0818
Mean squared error: 0.0668
Mean squared error: 0.1260
Mean squared error: 0.1036
Mean squared error: 0.1147
Mean squared error: 0.1756
Mean squared error: 0.1699
Mean squared error: 0.1796
Mean squared error: 0.1728


Explaination : Mean Squared Error 
3-a-i) The mean squared error for 10 fold CV can be found above.

In [12]:
LinReg= linearRegression()
LinReg.fit(X,Y)
predictedY=LinReg.predict(X)
meanSquaredError(predictedY,Y)

Mean squared error: 0.1159


Explaination : Mean Squared Error on the training set
3-a-ii) The mean squared error on the training set can be found above (0.1159).

In [14]:
feature_indx = np.argsort(LinReg.coef_)[::-1]
for i in feature_indx[:10]:
    print (Features[i], LinReg.coef_[i])

population 3.92893068558
PersPerOccupHous 1.20533188999
medFamInc 0.926851094478
PctOccupMgmtProf 0.743965216916
PctRecImmig8 0.712338606902
RentHighQ 0.670068418017
MalePctDivorce 0.599112582753
PctRecImmig5 0.585686876654
PctHousOwnOcc 0.535955037315
PctForeignBorn 0.446307015915


Explaination : Features that predict high crime 
3-a-iii) The features that predict high crime can be found above. I think these features are apt and ideal to predict high Crime in a particular community  

In [16]:
feature_indx = np.argsort(LinReg.coef_)
for i in feature_indx[:10]:
    print (Features[i], LinReg.coef_[i])


numbUrban -2.83057872862
PctKids2Par -1.23646333742
PctRecImmig10 -1.2130652747
medIncome -1.01150012138
pctWInvInc -0.94625437365
NumIlleg -0.847099741612
TotalPctDiv -0.651321367827
MedRent -0.583415073933
PctRecentImmig -0.56940638336
PctBSorMore -0.492669766355


Explaination : Features that predict low crime 
3-a-iii) The features that predict low crime can be found above. I think these features are apt and ideal to predict low Crime in a particular community  